storage/buffer.go

```diff
@@ -166,56 +164,121 @@ func (b *BufferedSeriesIterator) Err() error {
 	return b.it.Err()
 }
 
-// TODO(beorn7): Consider having different sample types for different value types.
-type sample struct {
+type fSample struct {
+	t int64
+	f float64
+}
+
+func (s fSample) T() int64 {
+	return s.t
+}
+
+func (s fSample) F() float64 {
+	return s.f
+}
+
+func (s fSample) H() *histogram.Histogram {
+	panic("H() called for fSample")
+}
+
+func (s fSample) FH() *histogram.FloatHistogram {
+	panic("FH() called for fSample")
+}
```

Comment:
Should we return `nil` instead? Even for below, `nil`, or `0` in case of float, instead of panic. I don't see why we should panic.

Reply:
I would say that calling the "wrong" method on any of the `Sample` implementation is a programming error. If we returned `nil` or `0` in that case, we would hide that error, which could lead to hard to spot calculation errors (https://github.com/prometheus/prometheus/issues/11973 is somewhat in that category, although it was mostly known already).

Unless there is real use in returning `nil` or `0`, I think it's safer to stick with the panic.